CommandLineException is no longer silent #2324

lbergelson opened the pull request:
fixes the issue where CommandLineExceptions produced no error message

added a public getUsage() method to CommandLineProgram
added a catch for these in instanceMain(), where the CommandLineProgram is in scope for printing the usage message
added a protected accessor getCommandLineParser to CommandLineProgram which guards against having an uninitialized CommandLineParser
moved the "A USER ERROR HAS OCCURRED" text out of the actual user exception and into the pretty printing

magicDGS commented on src/main/java/org/broadinstitute/hellbender/Main.java:
```diff
@@ -117,24 +138,19 @@ protected final void mainEntry(final String[] args) {
-            if(printStackTraceOnUserExceptions()) {
-                e.printStackTrace();
-            }
+            handleUserException(e);
```
This will print also internal exceptions from Barclay, no?

lbergelson replied:
Yes, that's currently the case, but it's going to be fixed soon I believe, with those exceptions being moved to a different exception type.

cmnbroad replied:
Right - Barclay will be changed soon so that CommandLineException can always be treated as a user error, and internal exception will be separate. Once thats done I think this will do exactly the right thing.

magicDGS commented on src/main/java/org/broadinstitute/hellbender/Main.java:
```diff
@@ -93,7 +108,13 @@ public Object instanceMain(final String[] args, final List<String> packageList,
+            return program.instanceMain(mainArgs);
+        } catch (final CommandLineException e){
+            //Print usage in the case of a CommandLineException
+            System.err.println(program.getUsage());
```
Why not doing this in `mainEntry()`. I think that it is more configurable like that...

lbergelson replied:
I put it here because `mainEntry()` doesn't have access to the program itself, so it can't print the usage.  Will this cause problems for you?  I could pull up the call to `extractCommandLineProgram` into `mainEntry()` if having this here would cause you issues.

magicDGS replied:
If it is complicated, I think that it could be maintained here. I'm just thinking in putative usages in the future for me or other framework users...

magicDGS commented on src/main/java/org/broadinstitute/hellbender/Main.java:
```diff
@@ -58,6 +60,19 @@
+        Utils.nonNull(e, "exception");
+        ps.println("***********************************************************************");
+        ps.println();
+        ps.println("A USER ERROR has occurred: " + e.getMessage());
```
Maybe the preamble ("A USER ERROR has ocurred") should be a parameter to allow other decorated exceptions being printed in a different way?

cmnbroad replied:
Does this need to be public ?

magicDGS replied:
If someone extends main and want to handle some exceptions in a different way, decorating the exception may be useful as public or protected. And if the preamble is configurable, it may help to handle other exceptions differently for other toolkit.

lbergelson replied:
made private and extracted the a prefix argument

magicDGS replied:
I would rather prefer protected for usage in classes extending `Main`...

cmnbroad replied:
Protected is what I had in mind when originally made my comment. It just seemed unnecessary for it to be public.

lbergelson replied:
made protected

cmnbroad commented:
A couple of minor comments, otherwise LGTM.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineProgram.java:
```diff
@@ -204,34 +200,25 @@ public Object instanceMain(final String[] argv) {
     /**
     *
-    * @return true if command line is valid
+    * @return true if program should be executed, false if an information only argument like {@link SpecialArgumentsCollection#HELP_FULLNAME} was specified
```
I've changed the behavior of this method.  It now always throws if validation, including custom validation was unsuccessful.  I believe this is much more consistent behavior with the rest of the codebase.

lbergelson commented:
@cmnbroad Could you take a quick look at this again when you get a chance?  I changed a few things in the untested methods to respond to @magicDGS's comments, but since they're so important and basically untested I think it would be good for someone to scan them.  I change the behavior of failing customCommandLineValidation to always throw, so it's consistent failing regular command line validation as well.   I also fixed some comment formatting.

cmnbroad commented:
Sure.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/Main.java:
```diff
@@ -112,29 +136,28 @@ public Object instanceMain(final String[] args) {
-            if(printStackTraceOnUserExceptions()) {
-                e.printStackTrace();
+            if (program != null) { //program can be null if th
+                System.err.println(program.getUsage());
```
The comment trails off a bit at the end... Having said that I don't see a code path that gets here with program==null, since runCommandLineProgram just returns if it doesn't have a program. I suppose that could change in the future though.

lbergelson replied:
Hmn, that's a good point.  I was thinking  that since program could be null and we should guard against that case, but you're right that a null program should never throw a `CommandLineException`.   I'll drop it.
